### PR TITLE
Alerting: Move evaluation outside folder section, and move labels in instead

### PR DESCRIPTION
**Pull request by soniaAguilarPeiron**
**What is this feature?**

This PR moves elements in the alert rule form. In particular:
- Folder and labels are in the same section (organize rules)
- Evaluation : groups and evaluation fields are in a new same section (specific to evaluation)

**Why do we need this feature?**

We are improving the easy of use when creating a new alert rule.

**Who is this feature for?**

All alerting users.

**Which issue(s) does this PR fix?**:

part of https://github.com/grafana/alerting-squad/issues/931

**Special notes for your reviewer:**

you can test it in here => (https://ephemeral1511182195121soniaag.grafana-dev.net)

Screenshot 2024-11-07 at 09 42 29

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: public/app/features/alerting/unified/components/rules/EditRuleGroupModal.tsx [hunk 1]
```diff
@@ -178,7 +178,7 @@ export interface ModalProps {
   hideFolder?: boolean;
 }
 
-export function EditCloudGroupModal(props: ModalProps): React.ReactElement {
+export function EditRuleGroupModal(props: ModalProps): React.ReactElement {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/rules/EditRuleGroupModal.tsx:181]: I changed the name, as this group modal is not for cloud?🤔

File: public/app/features/alerting/unified/utils/rules.ts [hunk 2]
```diff
@@ -411,7 +411,7 @@ export function isGrafanaAlertingRuleByType(type?: RuleFormType) {
   return type === RuleFormType.grafana;
 }
 
-export function isGrafanaRecordingRuleByType(type: RuleFormType) {
+export function isGrafanaRecordingRuleByType(type?: RuleFormType) {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/utils/rules.ts:414]: added as optional to not have to check when the value is still undefined...

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      appEvents.emit(AppEvents.alertError, ['Folder could not be created']);
+      return;
+    }
+
+    const folder: Folder = { title: newFolder.title, uid: newFolder.uid };
+    onCreate(folder);
+    appEvents.emit(AppEvents.alertSuccess, ['Folder Created', 'OK']);
+  };
+
+  const error = containsSlashes(title);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:135]: I think slashes are no longer issue for GMA

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  onClose: () => void;
+  onCreate: (folder: Folder) => void;
+}): React.ReactElement {
+  const styles = useStyles2(getStyles);
+
+  const [title, setTitle] = useState('');
+  const onSubmit = async () => {
+    const newFolder = await createFolder({ title: title });
+    if (!newFolder.uid) {
+      appEvents.emit(AppEvents.alertError, ['Folder could not be created']);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:126]: Should we convert this to `useAppNotifications`?

**Comment by konrad147** [comment]
Question: Could separating folders and groups into different sections be confusing for users, since groups depend on folders? It looks like the design assumes evaluation groups will eventually be removed, which would make this layout clearer, but we’re not there yet.

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx (hunk 3 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:135]: good point, this is from old code moved here, but let's remove it 👍

**Comment by soniaAguilarPeiron** [comment]
> Question: Could separating folders and groups into different sections be confusing for users, since groups depend on folders? If you change the folder after selecting a group, we need to clear the group selection. It looks like the design assumes evaluation groups will eventually be removed, which would make this layout clearer, but we’re not there yet.

I understand what you mean, but I don't think users will get more confused... with the new UI , group selector is after folder selector and it's close to it. What I believe it could be confusing is changing this order, because groups depend on the folder selected as you mentioned before.

Screenshot 2024-11-07 at 09 35 40

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx (hunk 3 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:135]: I would remove this logic in another PR, wdyt?

**Comment by soniaAguilarPeiron** [comment]
I pushed a commit updating the label when group can not be selected yet @konrad147 

Screenshot 2024-11-07 at 10 36 02

File: public/app/features/alerting/unified/components/rule-editor/AnnotationsStep.tsx [hunk 5]
```diff
@@ -105,9 +107,11 @@ const AnnotationsStep = () => {
     );
   }
 
+  const step = !isGrafanaManagedRuleByType(type) ? 5 : 6;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/AnnotationsStep.tsx:110]: Can we add a comment here to capture the intent / business logic associated here?

**Comment by brendamuir** [comment]
couple of questions/comments -- 

1. so existing section 4 will become "Configure notifications"?
2. is there any other function of folders other than storing alert rules and then being able to give permissions to access those folders/alert rules contained in them?
3. We should call it "Add/select folder and labels" to be consistent with the other titles
4. We should add a text line below the title also to be consistent. - considering maybe a tooltip to clarify what folders are used for? eg. Labels are used to differentiate an alert from all other alerts.  You can use them for searching, silencing, and routing notifications. Folders are used for storing alert rules. You can extend the access provided by a role to alert rules and rule-specific silences by assigning permissions to individual folders.

**Comment by soniaAguilarPeiron** [comment]
> * so existing section 4 will become "Configure notifications"?

yes, but without labels

> * is there any other function of folders other than storing alert rules and then being able to give permissions to access those folders/alert rules contained in them?

I would say no
> * We should call it "Add/select folder and labels" to be consistent with the other titles

👍

> * We should add a text line below the title also to be consistent. - considering maybe a tooltip to clarify what folders are used for? eg. Labels are used to differentiate an alert from all other alerts.  You can use them for searching, silencing, and routing notifications. Folders are used for storing alert rules. You can extend the access provided by a role to alert rules and rule-specific silences by assigning permissions to individual folders.

💯

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+          </Field>
+        }
+        {isCreatingFolder && (
+          <FolderCreationModal onCreate={handleFolderCreation} onClose={() => setIsCreatingFolder(false)} />
+        )}
+      </Stack>
+
+      {isCreatingFolder && (
+        <FolderCreationModal onCreate={handleFolderCreation} onClose={() => setIsCreatingFolder(false)} />
+      )}
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:108]: opssss

File: public/locales/en-US/grafana.json [hunk 7]
```diff
@@ -275,6 +275,13 @@
         "new-folder": "New folder",
         "new-folder-or": "or"
       },
+      "folder-and-labels": "Organize your rule with a folder and a set of labels.",
+      "folders": {
+        "help-info": "Folders are used for storing alert rules. You can extend the access provided by a role to alert rules and assigng permissions to individual folders."
```
Comment (brendamuir) [inline public/locales/en-US/grafana.json:280]:
```suggestion
        "help-info": "Folders are used for storing alert rules. You can extend the access provided by a role to alert rules and assign permissions to individual folders."
```

File: public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+import { contextSrv } from 'app/core/services/context_srv';
+import { createFolder } from 'app/features/manage-dashboards/state/actions';
+import { AccessControlAction } from 'app/types';
+
+import { Trans } from '../../../../../core/internationalization/index';
+import { RuleFormValues } from '../../types/rule-form';
+
+import { containsSlashes, Folder, RuleFolderPicker } from './RuleFolderPicker';
+
+export function FolderWithoutGroup() {
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderWithoutGroup.tsx:19]: this component represents only the folder, it does not include labels.

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 9]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
+              data-testid={selectors.components.AlertRules.newEvaluationGroupButton}
+            >
+              <Trans i18nKey="alerting.rule-form.evaluation.new-group">New evaluation group</Trans>
+            </Button>
+          </Box>
+          {isCreatingEvaluationGroup && (
+            <EvaluationGroupCreationModal
+              onCreate={handleEvalGroupCreation}
+              onClose={() => setIsCreatingEvaluationGroup(false)}
+              groupfoldersForGrafana={groupfoldersForGrafana?.result}
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:342]: What happens if we display the modal but `groupfoldersForGrafana` is null? Should we prevent that?

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 10]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
 
-function ForInput({ evaluateEvery }: { evaluateEvery: string }) {
+function EvaluationGroupCreationModal({
+  onClose,
+  onCreate,
+  groupfoldersForGrafana,
+}: {
+  onClose: () => void;
+  onCreate: (group: string, evaluationInterval: string) => void;
+  groupfoldersForGrafana?: RulerRulesConfigDTO | null;
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:465]: Same question as above. Why do we want to allow null here?

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 11]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
-                      All rules in the selected group are evaluated every {{ evaluateEvery }}.
-                    </Trans>
-                    {!isNewGroup && (
-                      <IconButton
-                        name="pen"
-                        aria-label="Edit"
-                        disabled={editGroupDisabled}
-                        onClick={onOpenEditGroupModal}
-                      />
+    // TODO remove "and alert condition" for recording rules
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:268]: this is from previous code, but I believe we can safely remove it 👍

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 12]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
+              error={errors.group?.message}
+              invalid={!!errors.group?.message}
+              htmlFor="group"
+            >
+              <Controller
+                render={({ field: { ref, ...field }, fieldState }) => (
+                  <AsyncSelect
+                    disabled={!folder || loading}
+                    inputId="group"
+                    key={uniqueId()}
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:290]: this is from previous code, I only moved that code to another file/component

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx (hunk 9 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:342]: this is existing code (just moved)

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 13]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
+              onClose={() => setIsCreatingEvaluationGroup(false)}
+              groupfoldersForGrafana={groupfoldersForGrafana?.result}
+            />
+          )}
+        </Stack>
+
+        {folderName && isEditingGroup && (
+          <EditRuleGroupModal
+            namespace={existingNamespace ?? emptyNamespace}
+            group={existingGroup ?? emptyGroup}
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:350]: same

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx (hunk 10 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:465]: same

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:290]: Let's review it in a separate PR

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx [hunk 14]
```diff
@@ -114,60 +219,369 @@ function FolderGroupAndEvaluationInterval({
...
+                  <AsyncSelect
+                    disabled={!folder || loading}
+                    inputId="group"
+                    key={uniqueId()}
+                    {...field}
+                    onChange={(group) => {
+                      field.onChange(group.label ?? '');
+                    }}
+                    isLoading={loading}
+                    invalid={Boolean(folder) && !group && Boolean(fieldState.error)}
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:296]: Same

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:350]: this was added in this previous [commit](https://github.com/grafana/grafana/commit/69b2aade1bc77c9d2ba81342564ecb816c16bf9a) for some reason

File: public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaEvaluationBehavior.tsx:268]: actually, no, we should not remove it, I will open a separate PR with this change as we need to update some text specific for recording rules here

File: public/app/plugins/panel/nodeGraph/panelcfg.cue [hunk 15]
```diff
@@ -43,12 +43,12 @@ composableKinds: PanelCfg: {
 					// Unit for the secondary stat to override what ever is set in the data frame.
 					secondaryStatUnit?: string
 				}
-        ZoomMode: "cooperative" | "greedy" @cuetsy(kind="enum")
```
Comment (soniaAguilarPeiron) [inline public/app/plugins/panel/nodeGraph/panelcfg.cue:46]: these changes were applied after merging from main, probably a change in the format?

File: public/app/features/alerting/unified/components/rules/EditRuleGroupModal.tsx (hunk 1 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rules/EditRuleGroupModal.tsx:181]: Perhaps a left-over from some refactoring? :D

File: public/app/features/alerting/unified/components/rule-editor/GrafanaFolderAndLabelsStep.tsx [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+import { LabelsFieldInForm } from './labels/LabelsFieldInForm';
+
+/** Precondition: rule is Grafana managed.
+ */
+export function GrafanaFolderAndLabelsStep() {
+  const { setValue, getValues } = useFormContext<RuleFormValues>();
+  const [showLabelsEditor, setShowLabelsEditor] = useState(false);
+
+  function onCloseLabelsEditor(
+    labelsToUpdate?: Array<{
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/GrafanaFolderAndLabelsStep.tsx:23]: but I found the private type `KVObject` , so, let's make it public 👌

File: public/app/features/alerting/unified/components/rule-editor/FolderSelector.tsx [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+      notifyApp.error('Failed to create folder');
+    } else if (data) {
+      onCreate({ title: data.title, uid: data.uid });
+      notifyApp.success('Folder created');
+    }
+  };
+
+  return (
+    <Modal className={styles.modal} isOpen={true} title={'New folder'} onDismiss={onClose} onClickBackdrop={onClose}>
+      <div className={styles.modalTitle}>
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderSelector.tsx:141]: same, this is from previous code, this pr only moves it to another place

File: public/app/features/alerting/unified/components/rule-editor/FolderSelector.tsx [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+      <form onSubmit={onSubmit}>
+        <Field label={<Label htmlFor="folder">Folder name</Label>}>
+          <Input
+            data-testid={selectors.components.AlertRules.newFolderNameField}
+            autoFocus={true}
+            id="folderName"
+            placeholder="Enter a name"
+            value={title}
+            onChange={(e) => setTitle(e.currentTarget.value)}
+            className={styles.formInput}
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/FolderSelector.tsx:160]: same, old code, we as its an small change, we can do it 👍

**Review by gillesdemey — APPROVED** [review]
Code changes LGTM!

**Pull request merged**